File: Graphs/542. 01 Matrix/solution.cpp
class Solution {
public:
    vector<vector<int>> updateMatrix(vector<vector<int>>& mat) {
        int xdir[] = {0,0,1,-1};
        int ydir[] = {1,-1,0,0};
        int n = mat.size();
        int m = mat[0].size();
        vector<vector<int>> ls(n, vector<int>(m,0));
        vector<vector<int>> visited(n, vector<int>(m,0));
        queue<pair<int,int>> q;
        for(int i = 0; i < n; i++){
            for(int j = 0; j < m; j++){
                if(mat[i][j] == 0){
                    q.push({i,j});
                    visited[i][j] = 1;
                }
            }
        }
        int distance = 0;
        while(!q.empty()){
            int z = q.size();
            for(int i = 0; i < z; i++){
                pair<int, int> x = q.front();
                q.pop();
                ls[x.first][x.second] = distance;
                for(int j = 0; j < 4; j++){
                    int ni = x.first + xdir[j];
                    int nj = x.second + ydir[j];
                    if(ni >= 0 && ni < n && nj >= 0 && nj < m && visited[ni][nj] == 0){
                        q.push({ni,nj});
                        visited[ni][nj] = 1;
                    }
                }
            }

            distance++;
        }
        return ls;

        
    }
};

File: Graphs/542. 01 Matrix/solution.py
from queue import Queue
class Solution:
    def updateMatrix(self, mat: List[List[int]]) -> List[List[int]]:
        xdir = [0,0,1,-1]
        ydir = [1,-1,0,0]
        n = len(mat)
        m = len(mat[0])
        ls = [[0 for i in range(m)] for j in range(n)]
        visited = [[0 for i in range(m)] for j in range(n)]
        q = Queue()
        for i in range(n):
            for j in range(m):
                if mat[i][j] == 0:
                    q.put((i,j))
                    visited[i][j] = 1
        distance = 0
        while not q.empty():
            z = q.qsize()
            for i in range(z):
                x = q.get()
                ls[x[0]][x[1]] = distance
                for j in range(4):
                    ni = x[0] + xdir[j]
                    nj = x[1] + ydir[j]
                    if ni >= 0 and ni < n and nj >= 0 and nj < m and visited[ni][nj] == 0:
                        q.put((ni,nj))
                        visited[ni][nj] = 1
            distance += 1
        return ls
